### PR TITLE
chore: add events for timeouts

### PR DESCRIPTION
**Pull request by asalem1**
This PR adds eventing around query timeouts in the shared query API and notebooks query API. This PR also updates the metric selector bucket selection to group sample buckets, internal buckets, and user buckets for UI parity. Finally, this PR updates the way CSV chunking happens in notebooks to align with the rest of the UI and open the door for the eventually streaming of data as CSV partials since the previous implementation didn't give as granular control over the CSV chunking process

### CODE REVIEW

File: src/shared/contexts/query.tsx [hunk 1]
```diff
@@ -475,6 +538,13 @@ export const QueryProvider: FC = ({children}) => {
               const message = json.message || json.error
               const code = json.code
 
+              if (code === GATEWAY_TIMEOUT_STATUS) {
+                event('gateway timeout')
+              }
+              if (code === REQUEST_TIMEOUT_STATUS) {
+                event('query timeout')
+              }
```
Comment (drdelambre) [inline src/shared/contexts/query.tsx:546]: dont we also want a generic error case to be logged if it's not these two?

File: src/shared/contexts/query.tsx [hunk 2]
```diff
@@ -443,20 +479,47 @@ export const QueryProvider: FC = ({children}) => {
...
+            let didTruncate = false
+
+            let read = await reader.read()
+
+            while (!read.done) {
+              const text = decoder.decode(read.value)
+
+              bytesRead += read.value.byteLength
+
+              if (bytesRead > FLUX_RESPONSE_BYTES_LIMIT) {
```
Comment (drdelambre) [inline src/shared/contexts/query.tsx:502]: what does this mean?

File: src/shared/contexts/query.tsx [hunk 3]
```diff
@@ -443,20 +479,47 @@ export const QueryProvider: FC = ({children}) => {
...
+            const reader = response.body.getReader()
+            const decoder = new TextDecoder()
+
+            let csv = ''
+            let bytesRead = 0
+            let didTruncate = false
+
+            let read = await reader.read()
+
+            while (!read.done) {
```
[the remark this answers is not in the document]
Comment (drdelambre) [reply src/shared/contexts/query.tsx:497]: yeah

File: src/shared/contexts/query.tsx (hunk 2 above)
Comment (drdelambre) [reply src/shared/contexts/query.tsx:502]: can we remove it?

**Review by rockstar — APPROVED** [review]
As I read through this patch, I found myself nodding to @drdelambre's review notes. I leave how you handle those notes to your judgment.

This patch was a bit hard for me to review because it does three disparate things. It required a lot more cognitive load than I'm normally comfortable devoting to a review. It's more helpful for me-as-reviewer to see a single atomic change in PR form.

File: src/shared/contexts/query.tsx [hunk 4]
```diff
@@ -443,20 +448,39 @@ export const QueryProvider: FC = ({children}) => {
...
-                bytesRead: csv.length,
-                didTruncate: false,
+            let csv = ''
+            let bytesRead = 0
+
+            let read = await reader.read()
+
+            while (!read.done) {
+              if (!pending.current[id]) {
+                new CancellationError()
```
Comment (drdelambre) [inline src/shared/contexts/query.tsx:463]: does this throw it too?

**Review by drdelambre — APPROVED** [review]
good groundwork for partial csv handling!

**Pull request merged**